### PR TITLE
ADMINUI-2229 Want Filter on Services Page: Manta, SDC, or Both

### PR DESCRIPTION
**Pull request by Arc311**

### CODE REVIEW

File: www/js/views/services.js [hunk 1]
```diff
@@ -108,23 +123,43 @@ var ApplicationsListView = CollectionView.extend({
...
+        'click a[data-service-type]': 'onChangeView'
+    },
     name: 'services',
     url: function() {
         return '/services';
     },
 
     initialize: function() {
         this.applications = new Applications();
+        this.type = 'all';
```
Comment (andyYatskov) [inline www/js/views/services.js]: DEFAULT_TYPE = 'all';

File: www/js/views/services.js [hunk 2]
```diff
@@ -32,7 +33,15 @@ var Application = require('../models/model').extend({
 
 var Applications = require('../models/collection').extend({
     model: Application,
-    url: '/api/applications'
+    url: '/api/applications',
+    filterByType: function (type) {
+        if (type !== 'all') {
+            var applications = this.filter(function (application) {
```
Comment (andyYatskov) [inline www/js/views/services.js]: define variable before 'if'

File: www/js/views/services.js [hunk 3]
```diff
@@ -140,7 +141,7 @@ var View = Backbone.Marionette.ItemView.extend({
 
     initialize: function() {
         this.applications = new Applications();
-        this.type = 'all';
+        this.DEFAULT_TYPE = 'all';
```
Comment (andyYatskov) [inline www/js/views/services.js]: without this, make common variable

File: www/js/views/services.js [hunk 4]
```diff
@@ -35,8 +35,9 @@ var Applications = require('../models/collection').extend({
     model: Application,
     url: '/api/applications',
     filterByType: function (type) {
+        var applications = null;
         if (type !== 'all') {
```
Comment (andyYatskov) [inline www/js/views/services.js]: DEFAULT_TYPE

File: www/js/views/services.js [hunk 5]
```diff
@@ -10,6 +10,7 @@
 
 var $ = require('jquery');
 
+var _ = require('underscore');
 var Backbone = require('backbone');
 var adminui = require('../adminui');
 var Services = require('../models/services');
```
Comment (andyYatskov) [inline www/js/views/services.js]: this not need

Comment (andyYatskov) [inline www/js/views/services.js]: var Model = require('../models/model');  and replace below. also the same for collection

File: www/js/views/services.js [hunk 6]
```diff
@@ -108,23 +124,43 @@ var ApplicationsListView = CollectionView.extend({
...
     template: require('../tpl/services.hbs'),
+    events: {
+        'click a[data-service-type]': 'onChangeView'
+    },
     name: 'services',
     url: function() {
         return '/services';
     },
 
     initialize: function() {
```
Comment (andyYatskov) [inline www/js/views/services.js]: add space everywhere

File: www/js/views/services.js [hunk 7]
```diff
@@ -10,6 +10,7 @@
 
 var $ = require('jquery');
```
Comment (andyYatskov) [inline www/js/views/services.js]: remove line

File: www/js/views/services.js [hunk 8]
```diff
@@ -108,23 +124,43 @@ var ApplicationsListView = CollectionView.extend({
...
+            this.DEFAULT_TYPE = newType;
+            this.makeActive(newType);
+            this.applicationsView.collection = this.applications.filterByType(newType);
+            this.applicationsView.render();
+        }
     },
 
     onRender: function() {
         adminui.vent.trigger('settitle', 'services');
+        this.makeActive('all');
```
Comment (andyYatskov) [inline www/js/views/services.js]: DEFAULT_TYPE

File: www/js/views/services.js [hunk 9]
```diff
@@ -54,19 +55,19 @@ var InstanceView = Backbone.Marionette.ItemView.extend({
...
         data.vm = data.type === 'vm';
         data.agent = data.type === 'agent';
         return data;
     },
-    onRender: function() {
+    onRender: function () {
         var self = this;
         if (this.model.get('type') === 'vm') {
-            $.get('/api/vms/'+this.model.get('uuid')).done(function(res) {
+            $.get('/api/vms/'+this.model.get('uuid')).done(function (res) {
```
Comment (andyYatskov) [inline www/js/views/services.js]: space before and after '+'

File: www/js/views/services.js [hunk 10]
```diff
@@ -54,19 +55,19 @@ var InstanceView = Backbone.Marionette.ItemView.extend({
...
                 self.$('.state').addClass(res.state).html(res.state);
                 self.$('.ram').addClass(res.state).html(res.ram + ' MB');
```
Comment (andyYatskov) [inline www/js/views/services.js]: it looks strange, can we use variables in template?

File: www/js/views/services.js [hunk 11]
```diff
@@ -80,16 +81,16 @@ var InstancesView = CollectionView.extend({
 
 var ServicesListItemView = Backbone.Marionette.ItemView.extend({
     template: require('../tpl/services-item.hbs'),
-    attributes: { 'class': 'service' },
-    initialize: function() {
+    attributes: {'class': 'service'},
```
Comment (andyYatskov) [inline www/js/views/services.js]: 'class' -> class

**Comment by andyYatskov** [comment]
🆗

**Comment by dan-luk** [comment]
👌

**Comment by dan-luk** [comment]
🔢

File: www/js/views/services.js [hunk 12]
```diff
@@ -9,33 +9,44 @@
...
-var Applications = require('../models/collection').extend({
+var Applications = Collection.extend({
     model: Application,
-    url: '/api/applications'
+    url: '/api/applications',
+    filterByType: function (type) {
+        var applications = null;
+        if (type !== DEFAULT_TYPE) {
+            applications = this.filter(function (application) {
+                return application.attributes.name.toLowerCase() === type;
```
Comment (askfongjojo) [inline www/js/views/services.js]: The literals comparison is syntactically incorrect here. Regardless, the 'type' field currently has either 'vm' or 'agent' as value and will not help to distinguish between sdc vs manta services. We'll need to add in some new code to first get all the application uuid for application name = sdc (and manta). Then use that uuid to filter for services with matching application_uuid.

**Comment by askfongjojo** [comment]
Turned out that my earlier in-line comments on the filtering were incorrect. The application filtering actually failed to take effect when the param was passed as "applications_uuid" (with an extra 's', in the original code prior to this change). I've fixed that and confirmed that the change is working in an environment with sdc and manta installed. The commit has been merged as 8429f22.

**Comment by dan-luk** [comment]
thank you @askfongjojo

**Pull request closed without merging**